File: src/dsp/pfb_clock_sync.cpp
#include "pfb_clock_sync.h"

#include <algorithm>
#include <cmath>

#include "filter_design.h"

namespace jalert::dsp {

PfbClockSync::LinkedListBuf::LinkedListBuf(int initial_capacity)
    : data_(std::max(initial_capacity, 16)) {}

Complex32 PfbClockSync::LinkedListBuf::operator[](int idx) const {
    const int n = static_cast<int>(data_.size());
    return data_[(start_ + idx) % n];
}

void PfbClockSync::LinkedListBuf::push_back(Complex32 z) {
    const int n = static_cast<int>(data_.size());
    if (count_ == n) grow();
    const int slot = (start_ + count_) % static_cast<int>(data_.size());
    data_[slot] = z;
    ++count_;
}

void PfbClockSync::LinkedListBuf::pop_front() {
    if (count_ == 0) return;
    start_ = (start_ + 1) % static_cast<int>(data_.size());
    --count_;
}

void PfbClockSync::LinkedListBuf::grow() {
    const int old_n = static_cast<int>(data_.size());
    std::vector<Complex32> bigger(old_n * 2);
    for (int i = 0; i < count_; ++i) bigger[i] = data_[(start_ + i) % old_n];
    data_ = std::move(bigger);
    start_ = 0;
}

PfbClockSync::PfbClockSync(int nfilt, float sps_actual, float loop_bw, float max_dev,
                           int rrc_symbols, float rolloff)
    : nfilt_(nfilt),
      sps_(static_cast<int>(std::max(std::floor(sps_actual), 1.0f))),
      buf_(16) {
    constexpr int kPfbMasterSymbols = 8;
    const int total_taps = nfilt * std::max(rrc_symbols, kPfbMasterSymbols * sps_);
    auto master = rrc_taps(static_cast<float>(nfilt),
                            static_cast<float>(nfilt * sps_),
                            1.0f, rolloff, total_taps);
    arms_ = polyphase_decompose(master, nfilt);
    auto dtaps = create_diff_taps(master, nfilt);
    diff_arms_ = polyphase_decompose(dtaps, nfilt);
    taps_per_arm_ = static_cast<int>(arms_[0].size());

    // Critically damped: damping = 2 · nfilt; α/β derived from loop_bw.
    const float damping = 2.0f * nfilt;
    const float denom = 1.0f + 2.0f * damping * loop_bw + loop_bw * loop_bw;
    alpha_ = 4.0f * damping * loop_bw / denom;
    beta_  = 4.0f * loop_bw * loop_bw / denom;
    max_dev_ = max_dev;

    const float frac_sps = sps_actual - sps_;
    const float total = frac_sps * nfilt;
    rate_i_ = std::floor(total);
    rate_f_ = (total - rate_i_) / static_cast<float>(sps_ + 1);
    initial_rate_i_ = rate_i_;
    initial_rate_f_ = rate_f_;

    k_ = nfilt / 2.0f;
    prefix_keep_ = sps_ + 2;

    buf_ = LinkedListBuf(taps_per_arm_ + 4 * sps_);
}

bool PfbClockSync::step(Complex32 z, Complex32& output) {
    buf_.push_back(z);
    return try_emit(output);
}

void PfbClockSync::reset_loop() {
    k_ = nfilt_ / 2.0f;
    rate_f_ = initial_rate_f_;
    rate_i_ = initial_rate_i_;
    error_ = 0.0f;
}

bool PfbClockSync::try_emit(Complex32& output) {
    const int m = taps_per_arm_;
    const int prefix = prefix_keep_;

    if (!warmed_) {
        const int needed = prefix + m + sps_;
        if (buf_.count() < needed) return false;
        read_pos_ = prefix;
        warmed_ = true;
    }

    const float k_save = k_;
    int count = 0;
    int filtnum = static_cast<int>(std::floor(k_));
    while (filtnum >= nfilt_) { k_ -= nfilt_; filtnum -= nfilt_; ++count; }
    while (filtnum < 0)       { k_ += nfilt_; filtnum += nfilt_; --count; }

    const int start = read_pos_ + count;
    const int end = start + m;
    if (start < 0 || end > buf_.count()) {
        k_ = k_save;
        return false;
    }

    const auto& arm = arms_[filtnum];
    const auto& darm = diff_arms_[filtnum];
    float out_re = 0.0f, out_im = 0.0f;
    float diff_re = 0.0f, diff_im = 0.0f;
    for (int j = 0; j < m; ++j) {
        const Complex32 s = buf_[start + j];
        out_re  += s.re * arm[j];
        out_im  += s.im * arm[j];
        diff_re += s.re * darm[j];
        diff_im += s.im * darm[j];
    }

    k_ += rate_i_ + rate_f_;

    const float err = (out_re * diff_re + out_im * diff_im) * 0.5f;
    error_ = err;

    for (int i = 0; i < sps_; ++i) {
        rate_f_ += beta_ * err;
        k_ += rate_f_ + alpha_ * err;
    }
    if (rate_f_ > max_dev_) rate_f_ = max_dev_;
    else if (rate_f_ < -max_dev_) rate_f_ = -max_dev_;

    const int advance = count + sps_;
    if (advance >= 0) {
        read_pos_ += advance;
    } else {
        const int back = -advance;
        read_pos_ = read_pos_ > back ? read_pos_ - back : 0;
    }

    while (read_pos_ > prefix * 2) {
        buf_.pop_front();
        --read_pos_;
    }

    output = Complex32(out_re, out_im);
    return true;
}

} // namespace jalert::dsp

File: SDRSharp.JAlert/Dsp/PfbClockSync.cs
// Polyphase clock sync. Master RRC at nfilt × sps virtual rate, with a
// derivative arm set driving a Mueller-Müller-style timing-error detector. This
// block is also the chain's matched filter. The TED is modulation-agnostic and
// works for the J-ALERT BPSK. Ported from src/dsp/pfb_clock_sync.{h,cpp}.

using System;

namespace SDRSharp.JAlert.Dsp
{
    internal sealed class PfbClockSync
    {
        private sealed class LinkedListBuf
        {
            private Complex32[] _data;
            private int _start;
            private int _count;

            public LinkedListBuf(int initialCapacity)
            {
                _data = new Complex32[Math.Max(initialCapacity, 16)];
            }

            public int Count => _count;

            public Complex32 this[int idx]
            {
                get
                {
                    int n = _data.Length;
                    return _data[(_start + idx) % n];
                }
            }

            public void PushBack(Complex32 z)
            {
                int n = _data.Length;
                if (_count == n) Grow();
                int slot = (_start + _count) % _data.Length;
                _data[slot] = z;
                ++_count;
            }

            public void PopFront()
            {
                if (_count == 0) return;
                _start = (_start + 1) % _data.Length;
                --_count;
            }

            private void Grow()
            {
                int oldN = _data.Length;
                Complex32[] bigger = new Complex32[oldN * 2];
                for (int i = 0; i < _count; ++i) bigger[i] = _data[(_start + i) % oldN];
                _data = bigger;
                _start = 0;
            }
        }

        private readonly int _nfilt;
        private readonly int _sps;
        private readonly int _tapsPerArm;
        private readonly float[][] _arms;
        private readonly float[][] _diffArms;

        private readonly float _alpha;
        private readonly float _beta;
        private readonly float _maxDev;

        private float _rateF;
        private float _rateI;
        private float _k;
        private float _error;

        private readonly float _initialRateF;
        private readonly float _initialRateI;

        private LinkedListBuf _buf;
        private int _readPos;
        private readonly int _prefixKeep;
        private bool _warmed;

        public float Error => _error;
        public float RateF => _rateF;

        public PfbClockSync(int nfilt, float spsActual, float loopBw, float maxDev, int rrcSymbols, float rolloff)
        {
            _nfilt = nfilt;
            _sps = (int)Math.Max(Math.Floor(spsActual), 1.0f);
            _buf = new LinkedListBuf(16);

            const int kPfbMasterSymbols = 8;
            int totalTaps = nfilt * Math.Max(rrcSymbols, kPfbMasterSymbols * _sps);
            float[] master = FilterDesign.RrcTaps(nfilt, nfilt * _sps, 1.0f, rolloff, totalTaps);
            _arms = FilterDesign.PolyphaseDecompose(master, nfilt);
            float[] dtaps = FilterDesign.CreateDiffTaps(master, nfilt);
            _diffArms = FilterDesign.PolyphaseDecompose(dtaps, nfilt);
            _tapsPerArm = _arms[0].Length;

            // Critically damped: damping = 2 · nfilt; α/β derived from loop_bw.
            float damping = 2.0f * nfilt;
            float denom = 1.0f + 2.0f * damping * loopBw + loopBw * loopBw;
            _alpha = 4.0f * damping * loopBw / denom;
            _beta = 4.0f * loopBw * loopBw / denom;
            _maxDev = maxDev;

            float fracSps = spsActual - _sps;
            float total = fracSps * nfilt;
            _rateI = (float)Math.Floor(total);
            _rateF = (total - _rateI) / (_sps + 1);
            _initialRateI = _rateI;
            _initialRateF = _rateF;

            _k = nfilt / 2.0f;
            _prefixKeep = _sps + 2;

            _buf = new LinkedListBuf(_tapsPerArm + 4 * _sps);
        }

        // Push z; returns true and writes output when the loop emits a symbol.
        public bool Step(Complex32 z, out Complex32 output)
        {
            _buf.PushBack(z);
            return TryEmit(out output);
        }

        // Snap the PI-loop state back to the ctor defaults; leaves the sample ring.
        public void ResetLoop()
        {
            _k = _nfilt / 2.0f;
            _rateF = _initialRateF;
            _rateI = _initialRateI;
            _error = 0.0f;
        }

        private bool TryEmit(out Complex32 output)
        {
            output = default;
            int m = _tapsPerArm;
            int prefix = _prefixKeep;

            if (!_warmed)
            {
                int needed = prefix + m + _sps;
                if (_buf.Count < needed) return false;
                _readPos = prefix;
                _warmed = true;
            }

            float kSave = _k;
            int count = 0;
            int filtnum = (int)Math.Floor(_k);
            while (filtnum >= _nfilt) { _k -= _nfilt; filtnum -= _nfilt; ++count; }
            while (filtnum < 0) { _k += _nfilt; filtnum += _nfilt; --count; }

            int start = _readPos + count;
            int end = start + m;
            if (start < 0 || end > _buf.Count)
            {
                _k = kSave;
                return false;
            }

            float[] arm = _arms[filtnum];
            float[] darm = _diffArms[filtnum];
            float outRe = 0.0f, outIm = 0.0f;
            float diffRe = 0.0f, diffIm = 0.0f;
            for (int j = 0; j < m; ++j)
            {
                Complex32 s = _buf[start + j];
                outRe += s.Re * arm[j];
                outIm += s.Im * arm[j];
                diffRe += s.Re * darm[j];
                diffIm += s.Im * darm[j];
            }

            _k += _rateI + _rateF;

            float err = (outRe * diffRe + outIm * diffIm) * 0.5f;
            _error = err;

            for (int i = 0; i < _sps; ++i)
            {
                _rateF += _beta * err;
                _k += _rateF + _alpha * err;
            }
            if (_rateF > _maxDev) _rateF = _maxDev;
            else if (_rateF < -_maxDev) _rateF = -_maxDev;

            int advance = count + _sps;
            if (advance >= 0)
            {
                _readPos += advance;
            }
            else
            {
                int back = -advance;
                _readPos = _readPos > back ? _readPos - back : 0;
            }

            while (_readPos > prefix * 2)
            {
                _buf.PopFront();
                --_readPos;
            }

            output = new Complex32(outRe, outIm);
            return true;
        }
    }
}
